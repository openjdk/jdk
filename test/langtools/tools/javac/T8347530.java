/*
 * @test /nodynamiccopyright/
 * @bug 8347530
 * @summary Improve error message with invalid permits clauses
 * @compile/fail/ref=T8347530.out -XDrawDiagnostics T8347530.java
 */

class T8347530 {
    // sealed interfaces
    sealed interface A0 permits B0 {}
    class B0 {}     // class {0} must implement sealed interface

    sealed interface A1 permits B1 {}
    record B1() {}  // record {0} must implement sealed interface

    sealed interface A2 permits B2 {}
    enum B2 {}      // enum {0} must implement sealed interface

    sealed interface A3 permits B3 {}
    interface B3 {} // interface {0} must implement sealed interface

    // sealed classes
    sealed class C0 permits S0 {}
    class S0 {}     // class {0} must extend sealed class

    // record cannot extend other classes in general
    // enums cannot extend other classes in general
    // interfaces cannot extend other classes in general
}
