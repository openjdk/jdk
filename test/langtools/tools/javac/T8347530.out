T8347530.java:10:33: compiler.err.invalid.permits.clause: (compiler.misc.doesnt.implement.sealed: kindname.class, T8347530.B0)
T8347530.java:13:33: compiler.err.invalid.permits.clause: (compiler.misc.doesnt.implement.sealed: kindname.record, T8347530.B1)
T8347530.java:16:33: compiler.err.invalid.permits.clause: (compiler.misc.doesnt.implement.sealed: kindname.enum, T8347530.B2)
T8347530.java:19:33: compiler.err.invalid.permits.clause: (compiler.misc.doesnt.implement.sealed: kindname.interface, T8347530.B3)
T8347530.java:23:29: compiler.err.invalid.permits.clause: (compiler.misc.doesnt.extend.sealed: T8347530.S0)
5 errors
